# Exercise 15: Power Analyses

This  assignment is designed to give you practice with Monte Carlo methods to conduct power analyses via simulation. You won't need to load in any data for this homework. We will, however, be using parts of the homework from last week.

---
## 1. Simulating data (1 points)


Pull your `simulate_data()` function from your last homework and add it below. 

As a reminder, this function simulates the relationship between age, word reading experience, and reading comprehension skill. 

`c` is reading comprehension, and `x` is word reading experience.

In [1]:
sample_size = 100 # How many children in data set? 
age_lo = 80     # minimum age, in months
age_hi = 200    # maximum age, in months
beta_xa = 0.5   # amount by which experience changes for increase of one month in age
beta_x0 = -5    # amount of experience when age = 0 (not interpretable, since minimum age for this data is 80 months)
sd_x = 50       # standard dev of gaussian noise term, epsilon_x
beta_ca = 0.8   # amount that comprehension score improves for every increase of one unit in age
beta_cx = 3     # amount that comprehension score improves for every increase of one unit in reading experience
beta_c0 = 10    # comprehension score when reading experience is 0. 
sd_c = 85      # standard dev of gaussian noise term, epsilon_c

simulate_data <- function(sample_size, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c) {
      # WRITE YOUR CODE HERE
     
    r <- runif(sample_size, age_lo, age_hi)
    output_x = data.frame()
    output_c = data.frame()
    mean_r  <- mean(r) #generate mean(r)
    
    for (item in 1:sample_size)
        #x = (beta_xa + beta_x0 + rnorm(1, mean_x, sd_x))
        output_x = rbind(output_x, (beta_xa + beta_x0 + rnorm(1, mean_r, sd_x)))
    
    for (item in 1:sample_size)
        #x = (beta_xa + beta_x0 + rnorm(1, mean_x, sd_x))
        output_c = rbind(output_c, (beta_ca + beta_cx + beta_c0 + rnorm(1, mean_r, sd_c)))
    
    output = data.frame("age" = r, output_x, output_c)
    colnames(output)[2] ="x"
    colnames(output)[3] ="c"
    
    
      return(output)
}

dat <- simulate_data(sample_size, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c)
head(dat)

age       x        c        
1  91.27604 131.6742 132.42179
2  87.53383 210.1604 186.12311
3 158.92601 213.8303 227.93976
4 150.21775 216.1559  69.02675
5  98.22772 227.5536 172.14299
6 160.55970 157.5387  74.07795

---
## 2. `run_analysis()` function (2 pts)

Last week, we looked at whether word reading experience(`x`) mediated the relation between `age` and reading comprehension (`c`).

Now we're going to use our `simulate_data()` function to conduct a power analysis. The goal is to determine how many participants we would need in order to detect both the mediated and the direct effects in this data. 

*Note: We're going to pretend for the sake of simplicity that we don't have any control over the ages of the children we get (so ages are generated using `runif(sample_size, age_lo, age_hi)`, although of course this would be an unusual situation in reality.*

First, write a function, `run_analysis()`, that takes in simulated data, runs **your mediation from last week**, and returns a vector containing the ACME and ADE estimates and p-values (these are the `d0`, `d0.p`, `z0`, and `z0.p` features of the mediated model object, e.g., `fitMed$d0.p`). Print this function's output for the data we simulated previously. 

The following line confuses me, which data is it referring to?? :

Print this function's output for the data we simulated previously.

In [2]:
# WRITE YOUR CODE HERE
library(mediation)

#fitM <- lm(x ~ age,     data=dat) #Step 1: IV on M, age predicting x
#fitY <- lm(c ~ x + age, data=dat) #Step 2: IV and M on DV, x and age predicting c

#fitMed <- mediate(fitM, fitY, treat= "age", mediator= "c")
#summary(fitMed)

# a function for running a t-test and extracting p-value
run_analysis <- function(data) {
    fitM <- lm(x ~ age,    data=dat) #Step 1: IV on M, age predicting x
    fitY <- lm(c ~ x + age,    data=dat) #Step 2: IV and M on DV, x and age predicting c
    fit <- mediate(fitM, fitY, treat= "age", mediator= "c")  
    
    output  <- c(fit$d0, fit$d0.p, fit$z0, fit$z0.p)
    print(summary(fit))
    return(output)
    #return(summary(fit))
    #return(c(fit$estimate[1] - fit$estimate[2],fit$stderr,fit$p.value)) # this outputs the effect size, standard error, and significance for quality metrics 
}

rad  <- run_analysis(dat)
head(rad)

Warning message:
"package 'mediation' was built under R version 4.2.3"
Loading required package: MASS

Loading required package: Matrix

Loading required package: mvtnorm

Loading required package: sandwich

Warning message:
"package 'sandwich' was built under R version 4.2.3"
mediation: Causal Mediation Analysis
Version: 4.5.0





Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE               0.132       -0.262         0.52    0.54
Total Effect      0.132       -0.262         0.52    0.54
Prop. Mediated    0.000        0.000         0.00    1.00

Sample Size Used: 100 


Simulations: 1000 



[1] 0.0000000 1.0000000 0.1322787 0.5360000

---
## 3. `repeat_analysis()` function (3 pts)

Next fill in the function `repeat_analysis()` below so that it simulates and analyzes data `num_simulations` times. Store the outputs from each simulation in the `simouts` matrix. Calculate and return the coverage across all the simulations run for both ACME and ADE.

In [3]:
num_simulations = 1
alpha = .05

repeat_analysis <- function(num_simulations, alpha, sample_size, age_lo, age_hi, 
        beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c) {
    # Initialize simouts matrix for storing each output from run_analysis()
    simouts <- matrix(rep(NA, num_simulations*4), nrow=num_simulations, ncol=4)
    
        
    # Start simulating
    for (i in 1:num_simulations) {
      sim_dat  <- simulate_data(sample_size, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c)
      run_analysis  <- run_analysis(sim_dat)
      simouts[i, ]  <- run_analysis
    }
    
    # Calculate coverage for both ACME and ADE estimates using p-values in simouts
    #cvg <- mean(simouts[,3] <= alpha)
    ACME_cov = mean(simouts[,1]) # D0 = ACME
    ADE_cov = mean(simouts[,3]) # z0 = ADE

    return(list(ACME_cov = ACME_cov, ADE_cov = ADE_cov))
    #return(simouts)
}

repeat_analysis(num_simulations, alpha, sample_size, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c)
#head(repeat_analysis)


Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE               0.138       -0.281         0.55    0.52
Total Effect      0.138       -0.281         0.55    0.52
Prop. Mediated    0.000        0.000         0.00    1.00

Sample Size Used: 100 


Simulations: 1000 



$ACME_cov
[1] 0

$ADE_cov
[1] 0.1381998

In [4]:
simouts <- matrix(rep(NA, num_simulations*4), nrow=num_simulations, ncol=4)

simouts[1, ]  <- c(2, 3, 4, 4)
simouts

[,1] [,2] [,3] [,4]
[1,] 2    3    4    4

Now run the `repeat_analysis()` function using the same parameter settings as above, for 10 simulations, with an alpha criterion of 0.01. 

In [5]:
num_simulations = 10
alpha = .01

repeat_analysis(num_simulations, alpha, sample_size, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c)


Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE               0.137       -0.272         0.54    0.51
Total Effect      0.137       -0.272         0.54    0.51
Prop. Mediated    0.000        0.000         0.00    1.00

Sample Size Used: 100 


Simulations: 1000 


Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE               0.136       -0.269         0.54    0.52
Total Effect      0.136       -0.269         0.54    0.52
Prop. Mediated    0.000        0.000         0.00    1.00

Sample Size Used: 100 


Simulations: 1000 


Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE           

$ACME_cov
[1] 0

$ADE_cov
[1] 0.1342998

---
## 4. Testing different sample sizes (2 pts)

Finally, do the same thing (10 simulations, alpha criterion of 0.01) but for 5 different sample sizes: 50, 75, 100, 125, 150. You can do this using `map` (as in the tutorial), or a simple `for` loop, or by calculating each individually. Up to you! This should take around 3 minutes to run. 

In [20]:
# WRITE YOUR CODE HERE
library(purrr)

num_simulations = 2
alpha = .01

named_vector  <- c(1, 2, 3, 4)

sample_size = NA

for (item in named_vector)
    sample_size  <-  item
    print(item)
    repeat_analysis(num_simulations, alpha, item, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c)

[1] 4


Print your results. 

In [24]:
# WRITE YOUR CODE HERE
sample_size = 50

repeat_analysis(num_simulations, alpha, sample_size, age_lo, age_hi, beta_xa, beta_x0, sd_x, beta_ca, beta_cx, beta_c0, sd_c)


Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE               0.129       -0.281         0.53    0.53
Total Effect      0.129       -0.281         0.53    0.53
Prop. Mediated    0.000        0.000         0.00    1.00

Sample Size Used: 100 


Simulations: 1000 


Causal Mediation Analysis 

Quasi-Bayesian Confidence Intervals

               Estimate 95% CI Lower 95% CI Upper p-value
ACME              0.000        0.000         0.00    1.00
ADE               0.127       -0.271         0.53    0.57
Total Effect      0.127       -0.271         0.53    0.57
Prop. Mediated    0.000        0.000         0.00    1.00

Sample Size Used: 100 


Simulations: 1000 



$ACME_cov
[1] 0

$ADE_cov
[1] 0.1279429

## 5. Reflection (2 pts)

If this were a real power analysis, we'd want to run more simulations per sample size (to get a more precise estimate of power) and we may also want to test out some other values of the parameters we used to simulate our data. However, what would you conclude just based on the results above? 

> *Write your response here.* 
> 

**Given** how we generated the data, why was the direct effect harder to detect than the mediated effect?
> *Write your response here.* 

**DUE:** 5pm EST, April 5, 2023

**IMPORTANT** Did you collaborate with anyone on this assignment? If so, list their names here. 
> *Someone's Name*